cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -529,47 +541,143 @@ public Void call() throws Exception {
     });
   }
 
-  private void setInputDatasetIfNeeded(Job job) throws IOException {
-    InputFormatProvider provider = context.getInputFormatProvider();
+  // Multiple mappers could be defined, so we check that their output types are consistent.
+  // Return the TypeToken for one of the mappers (doesn't matter which one, since we check that all of their output
+  // key/value types are consistent).
+  // Returns null if their are no inputs configured.
+  @Nullable
+  private TypeToken<?> checkMapperOutputTypes(Job job) throws ClassNotFoundException {
+    int numInputs = context.getInputFormatProviders().size();
+    if (numInputs == 0) {
+      // if there are no inputs configured through CDAP APIs, nothing to check for consistency
+      return null;
+    }
+
+    List<Class<? extends Mapper>> mapperClasses = Lists.newArrayList(context.getMapperClasses().values());
 
-    // If input format is not set during beforeSubmit, use the one from the spec if it exists.
-    if (provider == null && specification.getInputDataSet() != null) {
-      context.setInput(specification.getInputDataSet());
-      provider = context.getInputFormatProvider();
+    if (mapperClasses.size() < numInputs) {
+      // means one of the inputs is going to default to use the mapper assigned on the job,
+      // so also check the job's mapper class
+      if (!isProgrammaticConfig(job.getConfiguration(), Job.MAP_CLASS_ATTR)) {
+        // mapper class was not explicitly set on the job (reduce only job)
+        return null;
+      }
+      mapperClasses.add(job.getMapperClass());
+    }
+
+    List<Class<?>> mapOutputKeys = new ArrayList<>();
+    List<Class<?>> mapOutputValues = new ArrayList<>();
+
+    TypeToken<?> type = null;
+    for (Class<? extends Mapper> mapperClass : mapperClasses) {
+      type = resolveClass(mapperClass, Mapper.class);
+      Type[] typeArgs = ((ParameterizedType) type.getType()).getActualTypeArguments();
+      mapOutputKeys.add(TypeToken.of(typeArgs[2]).getRawType());
+      mapOutputValues.add(TypeToken.of(typeArgs[3]).getRawType());
     }
 
-    if (provider != null) {
+    if (!elementsEqual(mapOutputKeys) || !elementsEqual(mapOutputValues)) {
+      throw new IllegalArgumentException("Type mismatch in output type of mappers: " + mapperClasses);
+    }
+
+    // this shouldn't ever happen, because we know there's at least 1 input configured
+    Preconditions.checkNotNull(type);
+    return type;
+  }
+
+  // returns true if all of the elements of the input list are equal
+  private boolean elementsEqual(List<Class<?>> classes) {
```

Comment:
If you requires all the classes have to be the same, why not use an identity set (`Collections.newSetFromMap(new IdentityHashMap<Class<?>, Boolean>())` to begin with and just check the size == 1?